### PR TITLE
File cache system

### PR DESCRIPTION
**Pull request by rafaelvieiras**
### Explanation of the changes, problem that they are intended to fix.

**tl;dr**
- Cache for external images. Save external images registered on EPG and return over API.
- Cache for m3u requests (reduced from seconds to milliseconds)
- M3U functions relocate for a better organization
- HOST constant to store HOST full address using an Express interceptor on all requests.
- Store DATABASE in constants to prevent use progress.env on another files.

I made this solution because commonly plex set host  to `***.plex.direct`, this is a external address. This make many issues when internet connection are down, or slow.

Players likes Emby, has a feature to show `<icon>` from xml EPG like a media poster, and on my connection I'm receiving many issues related with Plex connection.

Taking these issues, I decided to make a simple cache system. This is a good step to speedup UX and looking to turn off Plex, and on future to make others caches for others HTTP request.

I took the opportunity to cache m3u and made a little refatoration on m3u API!
I made a m3u-service.js to manager M3U logics.

### All Submissions:

* [x] I have read the code of conduct.
* [x] I am submitting to the correct base branch

### New features

* [x] I understand that the feature may not be accepted if it doesn't fit the upstream app's planned design direction. But that in this case I am encouraged to share this as an available modification other users can use if they want.

### CODE REVIEW

File: src/xmltv.js [hunk 1]
```diff
@@ -1,12 +1,15 @@
 const XMLWriter = require('xml-writer')
 const fs = require('fs')
+const CacheImageService = require('./cache-image-service');
+let dataBase = undefined;
```
Comment (TimeBomb) [inline src/xmltv.js:4]: Is this variable being used outside of the `WriteXMLTV` function? If not, would we be able to remove this line and use `db` in `WriteXMLTV`, keeping it scoped just to that function?

File: src/constants.js [hunk 2]
```diff
@@ -5,5 +5,7 @@ module.exports = {
     TVGUIDE_MAXIMUM_FLEX_DURATION : 6 * 60 * 60 * 1000,
     TOO_FREQUENT: 100,
 
-    VERSION_NAME: "1.3.0-prerelease"
+    VERSION_NAME: "1.3.0-prerelease",
+    HOST: undefined,
```
Comment (TimeBomb) [inline src/constants.js:9]: If `HOST` and `DATABASE` are `undefined` here and we're setting them later, they're not really constants in the same way that these other numbers and strings in this file, right?

File: src/services/m3u-service.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
+const CacheService = require('../cache-service');
+const constants = require('../constants');
+
+/**
+ * Manager and Generate M3U content
+ *
+ * @class M3uService
+ */
+class M3uService {
```
Comment (TimeBomb) [inline src/services/m3u-service.js:9]: Small nitpick: We name this `M3uService`, but we have a method named `buildM3UList`. Could we standardize on either `M3u` or `M3U`?

File: src/cache-service.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+     * `get` a File from cache folder
+     *
+     * @param {string} fullFilePath
+     * @returns {promise} `Resolve` with file content, `Reject` with false
+     * @memberof CacheService
+     */
+    getCache(fullFilePath) {
+        return new Promise((resolve, reject) => {
+            try {
+                fs.readFile(path.join(this.cachePath, fullFilePath), 'utf8', function (err,data) {
```
Comment (TimeBomb) [inline src/cache-service.js:50]: If I understand this correctly, we're calling the `getCache` method in the following places:
`CacheImageService.getImageFromCache`
`M3uService.buildM3UList`
`M3uService.buildHLSList`

I'd expect these methods (or the methods that consume these methods) to be called potentially numerous times while a user is managing their dizquetv channels. If I'm mistaken, let me know.

With the implementation here, we'd be hitting the file system every single time those functions are called.

We may be able to gain performance speed by string the cached values in memory - in a JS object here - whenever we want to retrieve the cache.

That may look like:
1) On dizquetv start, preload commonly accessed cache into memory (in this service)
2) On `getCache` being called, return the variable from memory rather than reading from a file every time
3) Whenever we `setCache`, update the variable stored in memory _and_ update the file storage.

The negative here being memory impact. I'm not certain how large we can expect the files to be, and it may or may not be large enough that storing in memory uses up a painfully high amount of resources. Might be worth weighing one versus the other if we haven't already, though.

**Review by TimeBomb — COMMENTED** [review]
Cool stuff, left a few thoughts.

Question: I saw you mentioned ` (reduced from seconds to milliseconds)` - this is exciting sounding. Do you hapen to have any benchmark numbers available for a `before` and `after` of this change?

**Comment by vexorian** [comment]
xteve has an epg cache kinda think it is overkill to make one in dizquetv. And it can actually be problematic. Remember that requests made to dizquetv might block delivery blocks to the stream. And another benefit for having the cache in xteve is that xteve has authentication for the editing stuff. So if you wanted to share dizqueTV you would end up having two caches. One for the security and one  in dizqueTV. In my case I have a much greater worry regarding storage usage than EPG latency. Also this sort of thing makes containerization harder.

So I'll really need this EPG cache to be optional before I can merge it.

Regarding m3u cache, this might actually be unnecessary. I am guessing you have lots of channels. But there is an existing channel cache that we could  have used for this. Though certainly I bet it doesn't hurt, and I guess that even when using the channel cache it would still have less latency than caching the m3u.  Memory usage shouldn't be a big deal in the case for the m3u. So ok it can enter the feature set.

There are also other comments I have with code specifics.

**Comment by vexorian** [comment]
Oh, in fact the m3u cache does have a problem. It's actually important that the host in the m3u file can change depending on who makes the request. My xteve uses a different IP for calling to dizqueTV than my TV. 

Workaround would be to cache a file in which the host is a placeholder, and you then replace the placeholder with the request's host name. This will add a little more latency but let's be honest, it's a string replace so that's an extra millisecond at most.  Or the cache could have different values for each host. The problem here is that if for some reason there really are too many hosts calling it, then it could memory leak, but that sounds unlikely.

This will also remove the need for that hack of using HOST in constant.js

File: src/cache-service.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+const constants = require('./constants');
+
+/**
+ * A File Cache controller for store and retrieve files from disk
+ *
+ * @class CacheService
+ */
+class CacheService {
+    constructor() {
+        this.cachePath = path.join(constants.DATABASE, 'cache');
```
Comment (vexorian) [inline src/cache-service.js:12]: Instead of doing this constants.DATABASE thing, you should make the db a parameter to the constructor.

File: src/xmltv.js [hunk 6]
```diff
@@ -1,12 +1,15 @@
...
 
 module.exports = { WriteXMLTV: WriteXMLTV, shutdown: shutdown }
 
 let isShutdown = false;
 let isWorking = false;
 
-async function WriteXMLTV(json, xmlSettings, throttle ) {
+async function WriteXMLTV(json, xmlSettings, throttle, db) {
```
Comment (vexorian) [inline src/xmltv.js:11]: Instantiate the CacheImageService outside all of this. And pass it to writeXMLTV function instead of passing db.

**Review by vexorian — CHANGES_REQUESTED** [review]

**Comment by rafaelvieiras** [comment]
> xteve already has an EPG Image cache for those who need it so it might be overkill to make one in dizquetv. And it can actually be problematic. Remember that requests made to dizquetv might block delivery blocks to the stream. And another benefit for having the cache in xteve is that xteve has authentication for the editing stuff. So if you wanted to share dizqueTV you would end up having two caches. One for the security and one in dizqueTV. In my case I have a much greater worry regarding storage usage than EPG latency. Also this sort of thing makes containerization harder.
> 
> Although I am not entirely sure if it is useful to have it in dizqueTV. I am not opposed to having this feature. But I'll really need this EPG cache to be optional before I can merge it.
> 
> Regarding m3u cache, this might actually be unnecessary. I am guessing you have lots of channels. But there is an existing channel cache that we could have used for this. Though certainly I bet it doesn't hurt, and I guess that even when using the channel cache it would still have less latency than caching the m3u. Memory usage shouldn't be a big deal in the case for the m3u. So ok it can enter the feature set.
> 
> There are also other comments I have with code specifics.

@vexorian 
I understand your point and is very relevant.
But you're talking about external application cache, this is very important, but not make part of dizquetv. :bowtie:
I would want to offer those cache solutions not for a BIG speed difference, but to save resources to use on transcode, considering that the user does not have a GPU, for example (It's my case, I don't have a GPU on my Unraid :worried: ).

I would like to work on this toggle function to enable and disable Cache, this to maintain the cache feature. What do you think?

About m3u cache, You're right again! :sweat_smile:
But consider having a location on the disk for the user to access m3u over file. Many people use m3u lists per file, not URL. But again, I can add a toggle to turn this on and off. What do you think?

**Comment by rafaelvieiras** [comment]
@vexorian @TimeBomb I would like to thank all requests for changes and especially for being receptive to my suggestions to increase the application!

I made many changes!
### Turn Cache Image and Cache Service in Singleton
With This, I can call diskdb direct from class,  with performance and only one instance of class and diskdb! :D
Now I store the cache on memory and file. If cache solicited is on memory, I return without access file system. 

### Settings Class
I made this Class to create/update/read generic configurations.
And I start 2 configurations, one on cache-image-service and other on cache-service.

### config.js
I made this file to replace process.env use on other parts of code.
For now, I use on `settings-service`,  `cache-image-sevice` and `cache-service`. But on the future I would to like to replace all places are using  `process.env.DATABASE` on a new PR, if you accept and like the idea.

The problem of use porgress.env on any file is because progress.env are an OS variable, not an application variable. And this problem is aggravated when I change the value of this OS variable from inside application. This generates many bugs, for example, if a change progress.env on index.js, I can't start any other class or file before this change. So, my application turn in a waterfall application.

### Create a Cache Config Tab on UI
This is simple, I created a button toggle for enable and disable image and m3u cache.
On the future I would like to change the name of this tab from `Cache` to `Others Settings`.

**Comment by vexorian** [comment]
Ok. There's still some stuff I'd like to be different regarding the code and the settings, I will change them myself.
